plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -1656,23 +1711,52 @@ bool producer_plugin_impl::remove_expired_trxs( const fc::time_point& deadline )
    size_t num_expired_other = 0;
    size_t orig_count = _unapplied_transactions.size();
    bool exhausted = !_unapplied_transactions.clear_expired( pending_block_time, deadline,
-                  [&num_expired_persistent, &num_expired_other, pbm = _pending_block_mode,
+                  [this, &num_expired_persistent, &num_expired_other, pbm = _pending_block_mode,
                    &chain, has_producers = !_producers.empty()]( const transaction_id_type& txid, trx_enum_type trx_type ) {
+
             if( trx_type == trx_enum_type::persisted ) {
                if( pbm == pending_block_mode::producing ) {
                   fc_dlog(_trx_failed_trace_log,
                            "[TRX_TRACE] Block ${block_num} for producer ${prod} is EXPIRING PERSISTED tx: ${txid}",
                            ("block_num", chain.head_block_num() + 1)("txid", txid)
                            ("prod", chain.is_building_block() ? chain.pending_block_producer() : name()) );
+
+                  if (_trx_trace_dump_log.is_enabled(fc::log_level::debug)) {
+                      // Dump transaction
+                      std::shared_ptr<transaction> transaction_ptr = std::make_shared<transaction>(this->_unapplied_transactions.get_trx(txid)->packed_trx()->get_transaction());
```

Comment:
Should not make a copy of the transaction just to log it. Worse case you can add another variant for `transaction`.

Reply:
I tried passing `transaction` instead of `shared_ptr<transaction>` to the interface `get_entire_trx_trace()`, which had a variant for `transaction`, but there was compiling error `error: no viable conversion from ‘const eosio::chain::transaction’ to ‘const std::variant<fc::exception_ptr, transaction_trace_ptr, transaction>’` 

Will try agin.

Reply:
That is likely because `transaction` constructor is explicit which is a good thing because you would not want to make a copy just so you can call this. At this point, I would give up on using a std::variant.

Reply:
I will try making three diff methods to replace current `get_entire_trx_trace()`. I thought about making it a template and implementing it in `chain_plugin.hpp` but this interface needs to call `make_resolver` that is defined in `chain_plugin.cpp'.

Reply:
Fixed.